File: python/neura_x/liquid_router.py
# ==========================================================
# Neura-X: Intelligence Without Limits.
# Copyright (c) 2026 Edusei Mikel Lisamba. All Rights Reserved.
#
# Liquid Router — Dynamic Sparsity Engine
# ==========================================================

"""
Liquid Router for Neura-X.
Dynamically determines which neurons are awake and which are asleep.
"""

import numpy as np
from typing import Optional
from neura_x.module import Module, Parameter


class RoutingStats:
    """Statistics for routing operations."""
    def __init__(self):
        self.total_routed = 0
        self.total_awake = 0
        self.total_asleep = 0
        self.sparsity_ratio = 0.0
        self.avg_awake_fraction = 0.0


class LiquidRouter(Module):
    """Dynamic sparsity router for Neura-X."""

    def __init__(
        self,
        num_neurons: int,
        awake_threshold: float = 0.05,
        temperature: float = 1.0,
        temperature_decay: float = 0.999,
    ):
        super().__init__()
        self.num_neurons = num_neurons
        self.awake_threshold = awake_threshold
        self.temperature = temperature
        self.temperature_decay = temperature_decay
        self._step = 0

        self.register_parameter(
            "gate_logits",
            Parameter(np.zeros(num_neurons, dtype=np.float32))
        )

        self._total_routed = 0
        self._total_awake = 0

    def _gumbel_softmax(self, logits: np.ndarray, hard: bool = False) -> np.ndarray:
        if hard or self.temperature < 0.01:
            return (logits > 0).astype(np.float32)

        gumbel_noise = -np.log(-np.log(np.random.rand(*logits.shape) + 1e-10) + 1e-10)
        noisy_logits = (logits + gumbel_noise) / self.temperature

        exp_logits = np.exp(noisy_logits - np.max(noisy_logits))
        soft_gates = exp_logits / exp_logits.sum()
        return soft_gates

    def route(self, x: np.ndarray, hard: bool = False) -> np.ndarray:
        logits = self._parameters["gate_logits"].data
        gates = self._gumbel_softmax(logits, hard=hard or not self.training)

        num_awake = max(1, int(self.num_neurons * self.awake_threshold))
        top_indices = np.argsort(gates)[-num_awake:]

        mask = np.zeros(self.num_neurons, dtype=np.float32)
        mask[top_indices] = 1.0

        self._total_routed += self.num_neurons
        self._total_awake += int(mask.sum())

        return mask

    def forward(self, x: np.ndarray) -> np.ndarray:
        mask = self.route(x, hard=not self.training)
        return x * mask

    def step(self):
        self._step += 1
        self.temperature *= self.temperature_decay
        self.temperature = max(self.temperature, 0.01)

    @property
    def sparsity(self) -> float:
        if self._total_routed == 0:
            return 0.0
        return 1.0 - (self._total_awake / self._total_routed)

    @property
    def effective_awake(self) -> int:
        return max(1, int(self.num_neurons * self.awake_threshold))

    def get_stats(self) -> RoutingStats:
        """Get routing statistics."""
        stats = RoutingStats()
        stats.total_routed = self._total_routed
        stats.total_awake = self._total_awake
        stats.total_asleep = self._total_routed - self._total_awake
        if self._total_routed > 0:
            stats.sparsity_ratio = 1.0 - (self._total_awake / self._total_routed)
            stats.avg_awake_fraction = self._total_awake / self._total_routed
        return stats

    def memory_savings(self) -> dict:
        total_params = self.num_neurons
        awake_params = self.effective_awake
        return {
            "total_neurons": total_params,
            "awake_neurons": awake_params,
            "asleep_neurons": total_params - awake_params,
            "sparsity_percent": (1 - awake_params / total_params) * 100,
            "memory_reduction_factor": total_params / max(awake_params, 1),
        }

    def reset_stats(self):
        self._total_routed = 0
        self._total_awake = 0

    def __repr__(self) -> str:
        savings = self.memory_savings()
        return (
            f"LiquidRouter("
            f"neurons={self.num_neurons}, "
            f"awake={self.effective_awake}, "
            f"sparsity={savings['sparsity_percent']:.1f}%, "
            f"temp={self.temperature:.3f})"
        )

File: core/cpp/router/liquid_router.cpp
// ==========================================================
// Neura-X: Intelligence Without Limits.
// Copyright (c) 2026 Edusei Mikel Lisamba. All Rights Reserved.
//
// Liquid Router — Implementation
// ==========================================================

#include "router.h"
#include <cmath>
#include <cstring>
#include <algorithm>
#include <numeric>
#include <random>

namespace neurax {
namespace router {

LiquidRouter::LiquidRouter(
    int num_neurons,
    float awake_threshold,
    float temperature,
    float temperature_decay)
    : num_neurons_(num_neurons)
    , awake_threshold_(awake_threshold)
    , temperature_(temperature)
    , temperature_decay_(temperature_decay)
    , step_count_(0)
    , total_routed_(0)
    , total_awake_(0)
{
    gate_logits_.resize(num_neurons_, 0.0f);
}

LiquidRouter::~LiquidRouter()
{
}

void LiquidRouter::gumbel_softmax(
    const float* logits, float temp,
    float* gates, int n, bool hard)
{
    if (hard || temp < 0.01f) {
        for (int i = 0; i < n; i++) {
            gates[i] = (logits[i] > 0.0f) ? 1.0f : 0.0f;
        }
        return;
    }

    static std::mt19937 rng(42);
    std::uniform_real_distribution<float> dist(0.0005f, 0.9995f);

    std::vector<float> noisy(n);
    for (int i = 0; i < n; i++) {
        float u = dist(rng);
        float gumbel = -std::log(-std::log(u));
        noisy[i] = (logits[i] + gumbel) / temp;
    }

    // Softmax
    float max_val = *std::max_element(noisy.begin(), noisy.end());
    float sum = 0.0f;
    for (int i = 0; i < n; i++) {
        gates[i] = std::exp(noisy[i] - max_val);
        sum += gates[i];
    }
    if (sum > 1e-10f) {
        for (int i = 0; i < n; i++) {
            gates[i] /= sum;
        }
    }
}

int LiquidRouter::route(
    const float* input, int input_size, uint8_t* gate_mask)
{
    if (!gate_mask || num_neurons_ <= 0) return 0;
    (void)input;
    (void)input_size;

    // Compute soft gates
    std::vector<float> soft_gates(num_neurons_);
    bool hard = (temperature_ < 0.01f);
    gumbel_softmax(gate_logits_.data(), temperature_,
                   soft_gates.data(), num_neurons_, hard);

    // Apply sparsity: keep only top awake_threshold fraction
    int num_awake = std::max(1,
        static_cast<int>(num_neurons_ * awake_threshold_));

    // Find indices of top-K gates
    std::vector<int> indices(num_neurons_);
    std::iota(indices.begin(), indices.end(), 0);
    std::partial_sort(
        indices.begin(),
        indices.begin() + num_awake,
        indices.end(),
        [&soft_gates](int a, int b) {
            return soft_gates[a] > soft_gates[b];
        }
    );

    // Build gate mask
    std::memset(gate_mask, 0, num_neurons_);
    for (int i = 0; i < num_awake; i++) {
        gate_mask[indices[i]] = 1;
    }

    // Update statistics
    total_routed_ += num_neurons_;
    total_awake_ += num_awake;

    return num_awake;
}

void LiquidRouter::step()
{
    step_count_++;
    temperature_ *= temperature_decay_;
    if (temperature_ < 0.01f) temperature_ = 0.01f;
}

void LiquidRouter::set_awake_threshold(float threshold)
{
    if (threshold > 0.0f && threshold <= 1.0f) {
        awake_threshold_ = threshold;
    }
}

RoutingStats LiquidRouter::get_stats() const
{
    RoutingStats stats;
    stats.total_routed = total_routed_;
    stats.total_awake = total_awake_;
    stats.total_asleep = total_routed_ - total_awake_;
    stats.sparsity_ratio = (total_routed_ > 0)
        ? 1.0 - static_cast<double>(total_awake_) / total_routed_
        : 0.0;
    stats.avg_awake_fraction = (total_routed_ > 0)
        ? static_cast<double>(total_awake_) / total_routed_
        : 0.0;
    return stats;
}

void LiquidRouter::reset_stats()
{
    total_routed_ = 0;
    total_awake_ = 0;
}

int LiquidRouter::expected_awake() const
{
    return std::max(1, static_cast<int>(num_neurons_ * awake_threshold_));
}

} // namespace router
} // namespace neurax